File: SWEA/D4/1210. ［S／W 문제해결 기본］ 2일차 － Ladder1/［S／W 문제해결 기본］ 2일차 － Ladder1.py
T = 10
# 여러개의 테스트 케이스가 주어지므로, 각각을 처리합니다.
for test_case in range(1, T + 1):
    # ///////////////////////////////////////////////////////////////////////////////////
    t_c = int(input())
    ground = []
    for _ in range(100):
        line = list(map(int, input().split()))
        ground.append(line)

    now_row = 99
    now_col = ground[-1].index(2)

    visited = [[False] * 100 for _ in range(100)]

    while True:
        visited[now_row][now_col] = True

        if now_row == 0:
            break

        if now_row == 99:
            now_row -= 1
            continue

        # 좌, 우 확인
        if now_col - 1 >= 0 and ground[now_row][now_col - 1] == 1:
            if visited[now_row][now_col - 1] == False:
                now_col -= 1
                continue
        if now_col + 1 <= 99 and ground[now_row][now_col + 1] == 1:
            if visited[now_row][now_col + 1] == False:
                now_col += 1
                continue

        # 좌, 우에 길 없는 경우 -> 위로 올라가기
        now_row -= 1


    answer = now_col
    print(f"#{t_c} {answer}")

    # ///////////////////////////////////////////////////////////////////////////////////

File: SWEA/D4/1210. ［S／W 문제해결 기본］ 2일차 － Ladder1/［S／W 문제해결 기본］ 2일차 － Ladder1.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Solution {
	static int[][] ground;
	static int answer;
	
	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		
		for (int tc = 1; tc <= 10; tc++) {
			st = new StringTokenizer(br.readLine());
			String _ = st.nextToken();
			ground = new int[100][100];
			
			int endCol = 0;
			
			for (int i = 0; i < 100; i++) {
				st = new StringTokenizer(br.readLine());
				for (int j = 0; j < 100; j++) {
					int num = Integer.parseInt(st.nextToken());
					if (num == 2) {
						endCol = j;
					}
					ground[i][j] = num;
				}
			}
			
			DFS(99, endCol, 'B');
			System.out.println("#" + tc + " " + answer);
		}
	}
	
	private static void DFS(int nowRow, int nowCol, char direction) {
		if (nowRow == 0) {
			answer = nowCol;
			return;
		}
		
		if (direction == 'B') { // 1. 아래에서 온 경우
			if (0 <= nowCol - 1 && nowCol - 1 <= 99 && ground[nowRow][nowCol - 1] == 1) {
				DFS(nowRow, nowCol - 1, 'R');
			} else if (0 <= nowCol + 1 && nowCol + 1 <= 99 && ground[nowRow][nowCol + 1] == 1) {
				DFS(nowRow, nowCol + 1, 'L');
			} else {
				DFS(nowRow - 1, nowCol, 'B');
			}
		} else if (direction == 'L') { // 2. 왼쪽에서 온 경우
			if (ground[nowRow - 1][nowCol] == 1) {
				DFS(nowRow - 1, nowCol, 'B');
			} else if (0 <= nowCol + 1 && nowCol + 1 <= 99 && ground[nowRow][nowCol + 1] == 1) {
				DFS(nowRow, nowCol + 1, 'L');
			}
		} else if (direction == 'R') { // 3. 오른쪽에서 온 경우
			if (ground[nowRow - 1][nowCol] == 1) {
				DFS(nowRow - 1, nowCol, 'B');
			} else if (0 <= nowCol - 1 && nowCol - 1 <= 99 && ground[nowRow][nowCol - 1] == 1) {
				DFS(nowRow, nowCol - 1, 'R');
			}
		}
	}
}
